### PR TITLE
Increase default grpc max message size

### LINKED ISSUE #2301: Add support to handle gRpc messages > 4MB
gRpc has a max message size is limit  of 4MB.  We need to add support handle larger messages and in the meantime provide proper error message for messages larger than 4MB

### LINKED ISSUE #2668: Functions V2 Runtime errors on F5 for a Functions project in Visual Studio
_From @bzhu94 on April 12, 2018 16:33_

Please provide a succinct description of the issue.

#### Repro steps

INSTALL STEPS
1. Clean machine: Win 8.1 x64 ENU

2. Install  15.7 Preview 3 VSC (Web + Azure workload) from : \\\cpvsbuild\drops\VS\d15.7\products\27602.00\bootstrappers 

3. Select Web + Azure workload

4. Install the VSIX: \\\vwdbuild01\Drops\WTA\AzureFunctions-Dev15-OOB.Nightly\Dev15\2018-04-04_b5536\Release\Signed\VSIX\enu\Microsoft.VisualStudio.Web.AzureFunctions.vsix

#### Expected behavior

This error message should not appear when the runtime is started.

#### Actual behavior

The following error shows up upon starting the runtime: 

Unable to configure java worker.Could not find JAVA_HOME app setting.

Could not configure language worker Java

attachments

_Copied from original issue: Azure/azure-webjobs-sdk#1661_

### PR DESCRIPTION
**Pull request by pragnagopa**
Increase default grpc max message size to 32MB for dynamic sku, 128MB for dedicated. Fixes #2301 , Fixes #2668

### CODE REVIEW

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 1]
```diff
@@ -71,6 +71,8 @@ public class ScriptHost : JobHost
         internal static readonly TimeSpan MaxFunctionTimeout = TimeSpan.FromMinutes(10);
         private static readonly Regex FunctionNameValidationRegex = new Regex(@"^[a-z][a-z0-9_\-]{0,127}$(?<!^host$)", RegexOptions.Compiled | RegexOptions.IgnoreCase);
         public static readonly string Version = GetAssemblyFileVersion(typeof(ScriptHost).Assembly);
+        private static readonly int DefaultMaxMessageLengthDynamicSku = 32 * 1024 * 1024;
+        private static readonly int DefaultMaxMessageLength = 128 * 1024 * 1024;
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Same as above

File: src/WebJobs.Script.Grpc/Server/GrpcServer.cs [hunk 2]
```diff
@@ -15,9 +15,9 @@ public class GrpcServer : IRpcServer, IDisposable
         private Server _server;
         private bool _disposed = false;
 
-        public GrpcServer(FunctionRpc.FunctionRpcBase serviceImpl)
+        public GrpcServer(FunctionRpc.FunctionRpcBase serviceImpl, int grpcMaxMessageLength = 32 * 1024 * 1024)
```
Comment (paulbatum) [inline src/WebJobs.Script.Grpc/Server/GrpcServer.cs]: Its not great that the constant value is duplicated here... Any reason why this param has to be optional?

File: src/WebJobs.Script/Config/ScriptHostConfiguration.cs [hunk 3]
```diff
@@ -86,6 +86,11 @@ public ScriptHostConfiguration()
         /// </summary>
         public TimeSpan? FunctionTimeout { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value for grpc_max_message_length.
+        /// </summary>
+        public int MaxMessageLength { get; set; }
```
Comment (paulbatum) [inline src/WebJobs.Script/Config/ScriptHostConfiguration.cs]: Can we put the unit into the name (MaxMessageLengthBytes)

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 4]
```diff
@@ -1554,6 +1560,32 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+                {
+                    var requestedGrpcMaxMessageLength = int.Parse((string)value);
+                    if (ScriptSettingsManager.Instance.IsDynamicSku)
+                    {
+                        string message = $"Cannot set {nameof(scriptConfig.MaxMessageLength)} on Consumption plan. Default MaxMessageLength: {DefaultMaxMessageLengthDynamicSku} will be used";
+                        logger?.LogWarning(message);
+                    }
+                    else
+                    {
+                        scriptConfig.MaxMessageLength = requestedGrpcMaxMessageLength * 1024 * 1024;
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: The multiplication here has me confused.. is the unit of the setting in the json file in MB? Lets make it explicit in the name as discussed above.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 5]
```diff
@@ -1554,6 +1560,32 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+                        if (scriptConfig.MaxMessageLength < 0)
+                        {
+                            // Current grpc max message limits
+                            string message = $"MaxMessageLength must be between 4MB and 2000MB.Default MaxMessageLength: {DefaultMaxMessageLength} will be used";
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: this message describes a lower limit and upper limit but the comparison doesn't seem to match...

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 6]
```diff
@@ -1554,6 +1560,32 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+                            logger?.LogWarning(message);
+                        }
+                        scriptConfig.MaxMessageLength = scriptConfig.MaxMessageLength > 0 ? scriptConfig.MaxMessageLength : DefaultMaxMessageLength;
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: It looks like you're making zero mean "default" here. If you want these semantics, maybe make it nullable and then this line can could change into `scriptConfig.MaxMessageLength = scriptConfig.MaxMessageLength ?? Default`.

Alternatively, consider a design where this method does not update `scriptConfig.MaxMessageLength` multiple times. Might be better to have a local variable that it updates as it goes and then finally sets `scriptConfig.MaxMessageLength` when its done.

**Comment by pragnagopa** [comment]
@paulbatum - please let me know if you have any further feedback

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 7]
```diff
@@ -1619,10 +1623,48 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+            if (languageWorkerSection != null)
+            {
+                if (languageWorkerSection.TryGetValue("maxMessageLength", out value))
+                {
+                    requestedGrpcMaxMessageLength = int.Parse((string)value);
+                    if (ScriptSettingsManager.Instance.IsDynamicSku)
+                    {
+                        string message = $"Cannot set {nameof(scriptConfig.MaxMessageLengthBytes)} on Consumption plan. Default MaxMessageLength: {DefaultMaxMessageLengthBytesDynamicSku} will be used";
+                        logger?.LogWarning(message);
+                        requestedGrpcMaxMessageLength = DefaultMaxMessageLengthBytesDynamicSku;
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: as discussed, this line wont get hit if they have a language worker section but no max size, and thus the default won't be set to the dynamic default. Add a test case for this and you should see the failure

**Comment by pragnagopa** [comment]
@paulbatum - Addressed CR and updated. Thanks!

**Pull request merged**